### PR TITLE
8267773: PhaseStringOpts::int_stringSize doesn't handle min_jint correctly

### PR DESCRIPTION
**Pull request by neliasso**
Hi,

PhaseStringOpts::int_stringSize doesn't handle min_jint correctly - it can't be negated to a positive number. I fix this adding a special case for this.

Added a test that was contributed by Adam who reported the bug. Will add as a contributor.

Please review,
Best regards,
Nils Eliasson

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267773](https://bugs.openjdk.java.net/browse/JDK-8267773): PhaseStringOpts::int_stringSize doesn't handle min_jint correctly

### Reviewers
 * [Roland Westrelin](https://openjdk.java.net/census#roland) (@rwestrel - **Reviewer**) ⚠️ Review applies to da17bb3675114f104c55c2360bd45d331bd6ea22

### Contributors
 * Adam Farley `<afarley@openjdk.org>`
 * Aleksey Shipilev `<shade@openjdk.org>`
 * Nils Eliasson `<neliasso@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4255/head:pull/4255` \
`$ git checkout pull/4255`

Update a local copy of the PR: \
`$ git checkout pull/4255` \
`$ git pull https://git.openjdk.java.net/jdk pull/4255/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4255`

View PR using the GUI difftool: \
`$ git pr show -t 4255`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4255.diff">https://git.openjdk.java.net/jdk/pull/4255.diff</a>

</details>

### CODE REVIEW

**Review by rwestrel — APPROVED** [review]
Looks ok to me.

File: test/hotspot/jtreg/compiler/stringopts/TestFetchStaticField.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    public void runTest() {
+        System.out.println ("Test: StringIndexOutOfBoundsException(Integer.MIN_VALUE)");
+        for (int i = 0; i < 100_000; i++ ) {
+            run1Test(i);
+        }
+        System.out.println("Test finished.");
+    }
+
+    public static void main(String[] argv)
+    {
```
Comment (shipilev) [inline test/hotspot/jtreg/compiler/stringopts/TestFetchStaticField.java:48]: Inconsistent bracing style, should be on the same line?

File: test/hotspot/jtreg/compiler/stringopts/TestFetchStaticField.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+ * @library /test/lib /
+ * @run main compiler.stringopts.TestFetchStaticField
+ */
+
+package compiler.stringopts;
+
+import java.lang.StringIndexOutOfBoundsException;
+import jdk.test.lib.Asserts;
+
+public class TestFetchStaticField {
```
Comment (shipilev) [inline test/hotspot/jtreg/compiler/stringopts/TestFetchStaticField.java:37]: Why `TestFetchStaticField`? Is this really specific to static fields?

**Review by shipilev — COMMENTED** [review]
Huh, I found a similar issue with fuzzer tests this weekend. I minimized this to string concat bug as well, here is the test that I have: https://cr.openjdk.java.net/~shade/scratch/8267773.patch -- feel free to merge it here.

File: test/hotspot/jtreg/compiler/stringopts/TestFetchStaticField.java (hunk 2 above)
Comment (neliasso) [reply test/hotspot/jtreg/compiler/stringopts/TestFetchStaticField.java]: No - it's a leftover from when I was on the completely wrong track :P

I removed this test and added yours instead.

**Comment by neliasso** [comment]
> Looks ok to me.

Thank you Roland!

**Comment by shipilev** [comment]
My OpenJDK handle is "shade".

**Review by TobiHartmann — APPROVED** [review]
Looks good to me, thanks for fixing!

**Review by TobiHartmann — COMMENTED** [review]
What about the "wrong type in PhaseStringOpts::fetch_static_field" you mentioned in the bug comments?

**Comment by neliasso** [comment]
> What about the "wrong type in PhaseStringOpts::fetch_static_field" you mentioned in the bug comments?

It is a left over from a detour. Removed it.

**Pull request closed without merging**